File: src/main/java/com/brandonitaly/bedrockskins/mixins/MixinTranslationStorage.java
package com.brandonitaly.bedrockskins.mixins;

import com.brandonitaly.bedrockskins.pack.SkinPackLoader;
import net.minecraft.client.resources.language.ClientLanguage;
import org.spongepowered.asm.mixin.Mixin;
import org.spongepowered.asm.mixin.injection.At;
import org.spongepowered.asm.mixin.injection.Inject;
import org.spongepowered.asm.mixin.injection.callback.CallbackInfoReturnable;

@Mixin(ClientLanguage.class)
public class MixinTranslationStorage {
    @Inject(method = "getOrDefault", at = @At("HEAD"), cancellable = true)
    private void onGet(String key, String fallback, CallbackInfoReturnable<String> cir) {
        String custom = SkinPackLoader.getTranslation(key);
        if (custom != null) {
            cir.setReturnValue(custom);
        }
    }
}

File: src/main/kotlin/com/brandonitaly/bedrockskins/mixins/MixinTranslationStorage.kt
package com.brandonitaly.bedrockskins.mixins

import com.brandonitaly.bedrockskins.pack.SkinPackLoader
import net.minecraft.client.resource.language.TranslationStorage
import org.spongepowered.asm.mixin.Mixin
import org.spongepowered.asm.mixin.injection.At
import org.spongepowered.asm.mixin.injection.Inject
import org.spongepowered.asm.mixin.injection.callback.CallbackInfoReturnable

@Mixin(TranslationStorage::class)
class MixinTranslationStorage {
    @Inject(method = ["get"], at = [At("HEAD")], cancellable = true)
    private fun onGet(key: String, fallback: String, cir: CallbackInfoReturnable<String>) {
        val custom = SkinPackLoader.getTranslation(key)
        if (custom != null) {
            cir.returnValue = custom
        }
    }
}
